rs/replicated_state/src/page_map/storage.rs

```diff
@@ -152,14 +152,90 @@ impl BaseFile {
 /// from the newest overlay containing the page.
 /// The contents of pages that appear in no overlay file are read from `base`.
 #[derive(Clone, Default)]
-pub(crate) struct Storage {
+pub(crate) struct StorageImpl {
     /// The lowest level data we mmap during loading.
     base: BaseFile,
     /// Stack of overlay files, newest file last.
     overlays: Vec<OverlayFile>,
 }
 
+pub fn verify(storage_layout: &dyn StorageLayout) -> Result<(), PersistenceError> {
+    StorageImpl::load(storage_layout)?;
+    Ok(())
+}
+
+/// Lazy loaded representation of StorageImpl (see above).
+/// The `storage_layout` points to the files on disk, which are loaded at the first access.
+/// The loaded `StorageImpl` is never modified or unloaded till `drop`, meaning we don't read
+/// `storage_layout` ever again.
+/// If `storage_layout` is None during load we construct `StorageLayout` with the default
+/// constructor.
+#[derive(Default, Clone)]
+pub(crate) struct Storage {
+    storage_layout: Option<Arc<dyn StorageLayout + Send + Sync>>,
+    storage_impl: OnceLock<StorageImpl>,
+}
+
 impl Storage {
+    fn init_or_die(&self) -> &StorageImpl {
+        self.storage_impl
+            .get_or_init(|| match self.storage_layout.as_ref() {
+                None => Default::default(),
+                Some(storage_layout) => StorageImpl::load(storage_layout.deref())
+                    .expect("Failed to load storage layout"),
+            })
+    }
+
+    /// Whether the `storage_impl` is already loaded.
+    pub fn is_loaded(&self) -> bool {
+        self.storage_impl.get().is_some()
+    }
+
+    /// Create Storage.
+    pub fn lazy_load(
+        storage_layout: Arc<dyn StorageLayout + Send + Sync>,
+    ) -> Result<Self, PersistenceError> {
+        Ok(Storage {
+            storage_layout: Some(storage_layout),
+            storage_impl: OnceLock::default(),
+        })
+    }
+
+    pub fn get_page(&self, page_index: PageIndex) -> &PageBytes {
+        self.init_or_die().get_page(page_index)
+    }
+
+    pub fn get_base_memory_instructions(&self) -> MemoryInstructions {
+        self.init_or_die().get_base_memory_instructions()
+    }
+
+    pub fn get_memory_instructions(
+        &self,
+        range: Range<PageIndex>,
+        filter: &mut BitVec,
+    ) -> MemoryInstructions {
+        self.init_or_die().get_memory_instructions(range, filter)
+    }
+
+    pub fn num_logical_pages(&self) -> usize {
+        self.init_or_die().num_logical_pages()
+    }
+
+    pub fn serialize(&self) -> StorageSerialization {
+        self.init_or_die().serialize()
+    }
+
+    pub fn deserialize(serialized_storage: StorageSerialization) -> Result<Self, PersistenceError> {
+        let storage_impl = OnceLock::new();
+        let _ = storage_impl.set(StorageImpl::deserialize(serialized_storage)?);
+        Ok(Self {
+            storage_layout: None,
```

Comment:
We set `storage_layout` to `None`  after setting a value for `storage_impl` here. What about doing similar things in `init_or_die` or it is not that important?